### ISSUE TITLE
Rule Idea Board

### ISSUE THREAD

**Issue by DonIsaac** [DonIsaac/zlint#3]
This is a mega-issue tracking the list of rules we want to implement. Rule requests are welcome.

1. `no-undefined` - ban `undefined` in assignments and initializations
  - Allow when a `// SAFETY:` comment is present
  - Allow `self.* = undefined` in `deinit()` functions
 
2. `no-optional-unwrap`: bad optional unwrapping without a prior null check
3. `use-after-deinit`: ban references to `foo` after `foo.deinit()` calls (or other free-like calls)
4. `no-const-reassign`: do not reassign `const`-declared variables. Yes the compiler checks this, no they don't check it for tree-shaken code.
5. `no-private-access`: ban external references to `_private` container fields.

**Comment by DonIsaac** [comment]
I don't know what to call this, but @paperdave proposed it [here](https://discord.com/channels/876711213126520882/876711213126520885/1302145269189574706)
![image](https://github.com/user-attachments/assets/4e0dab6b-590c-4afc-b525-1b49ec59bcdd)

**Comment by DonIsaac** [comment]
Here's his full list (it needs triage)
![image](https://github.com/user-attachments/assets/ca3f993e-ce5f-41c0-bdb3-b7a4a668c4fe)

**Comment by paperdave** [comment]
> I don't know what to call this
"declaration can be accessed with a shorter path" / `shorter_decl_path`

**Comment by DonIsaac** [comment]
@paperdave `decl-alias`?

**Comment by DonIsaac** [comment]
`unused-array-side-effects`: Accidental mutation on a copied array
```zig
fn addEntry(self: *Foo, id: u32, item: u32) !void {
  // this creates a copy. It's dropped when the function returns
  var foos = self.foo[id]; // should be &self.foo[id];
  try foos.append(item);   // has no lasting side effects b/c foos is dropped
}
```

**Comment by MichaelBelousov** [comment]
`unsafe-union-unwrap`/`unchecked-union-variant/payload-reference`:

like `unsafe-option-unwrap` but enforces that a direct union variant access occurs in a scope where the union tag was checked and matches.

```zig
var a: union(enum) { int: i32, float: f32 } = .X;

if (a == .float) {
  std.debug.print("float: {}", .{a.float}); // legal
} else {
  std.debug.print("int: {}", .{a.float}); // illegal!
}

switch (a) {
  .int => std.debug.print("int: {}", .{a.float}), // illegal
}
```

**Comment by paperdave** [comment]
`naming-convention`: enforce standard library naming conventions declarations

**Comment by MichaelBelousov** [comment]
> it's better to avoid this pattern entirely, if possible. these rules should encourage `switch` and captures

I agree in general, but it's the same reason to have `unsafe-option-unwrap` even when `if (optional) |capture|` exists.

There is code when you're checking multiple conditions where a switch may be cumbersome:

```zig
if (condition && a == .float) {
  std.debug.print("{}", .{a.float});
}

vs:
if (condition) switch (a) { .float => |v| {
  std.debug.print("{}", .{v});
} }
```

**Comment by DonIsaac** [comment]
> > but enforces that a direct union variant access occurs in a scope where the union tag was checked and matches.
> 
> 
> 
> it's better to avoid this pattern entirely, if possible. these rules should encourage `switch` and captures:
> 
> 
> 
> ```zig
> 
> var a: union(enum) { int: i32, float: f32 } = .X;
> 
> 
> 
> if (a == .float) {
> 
>   std.debug.print("float: {}", .{a.float}); // illegal
> 
> } else {
> 
>   std.debug.print("int: {}", .{a.float}); // illegal
> 
> }
> 
> 
> 
> // rewrite into
> 
> switch (a) {
> 
>   .float => |float| std.debug.print("float: {}", .{a.float}),
> 
>   .int => |int| std.debug.print("int: {}", .{a.int}),
> 
> }
> 
> ```

Should we add a rule banning it entirely?

**Comment by MichaelBelousov** [comment]
> Should we add a rule banning it entirely?

Sure but I'd recommend allowing it with checks as the default.

It comes up often enough imo. A better example that I should have shown earlier is try checking that two separate variables have specific tags. You'd need a nested double switch if you don't allow tags checking in a single if statement

try rewriting this as a switch: 

```zig
if (a == .float and b == .int) {
}
```

**Comment by paperdave** [comment]
i think we should split these into separate issues. this thread is getting very long. one issue per lint rule, or for related ones they can share one.

**Comment by nektro** [comment]
https://github.com/nektro/ziglint/issues has a bunch i havent implemented

**Comment by michaelbartnett** [comment]
Variant on unsafe-optional-unwrap:

Ban `nullable.?`, require people spell out `nullable orelse unreachable` if they aren't going to properly do a safe if-with-capture or `nullable orelse <valid default expr>`.

Though reading your description of it, if you're talking about doing a typescript/c# style analysis where you're able to confirm a prior null check passed, then allowing `.?` if that is satisfied would be pretty nice. Not sure that's in scope for a linter though.

**Comment by felipeasimos** [comment]
I just opened a draft PR to implement a line length check rule (#232 ). I just realized I jumped the gun an didn't bring the idea here first (sorry).

The idea is to just check if any line surpasses a given threshold. Currently the only thing left to implement is to let the user define this threshold through the config (for now 120 is hardcoded).

Is this rule a good fit for the project?

**Comment by felipeasimos** [comment]
Would checking for empty `.zig` files be a valid idea? I'm not sure how zlint discover files. If its through imports it wouldn't really make sense. However, I have seen projects where empty `.zig` files were left in the code so it would be useful

**Comment by DonIsaac** [comment]
> Would checking for empty `.zig` files be a valid idea? I'm not sure how zlint discover files. If its through imports it wouldn't really make sense. However, I have seen projects where empty `.zig` files were left in the code so it would be useful

Yes

**Comment by felipeasimos** [comment]
From time to time I stumble in some comments in the zig community regarding `anytype` negatively. One of the most valid points being that it doesn't make the intent behind the code explicit, which kinda goes against the "easy to read" and "no hidden flow" objectives in the language itself. Maybe some people would find it useful having a `no-anytype` rule?

**Comment by damster101** [comment]
> the convention in zig is to write function names in camelCase and variables in snake_case. It would be cool to have a rule that checks at least if your functions are indeed camelCase. Coming from the rust world, where functions are also snake_case, this is a common problem

And type names in PascalCase.

**Comment by shakeelmohamed** [comment]
Hello, thanks for your effort on this project!

I’m curious if there’s interest in rules for alpha sorting imports. The eslint equivalent is [`sort-imports`](https://eslint.org/docs/latest/rules/sort-imports).

I found a file with 23 lines of imports of both kinds mixed in, and it looked quite bad (though a refactor is probably needed there anyways).  I’ve isolated 3 logical changes I would make if I was rewriting that blob:

## Sorting by package name

Before

```zig
const std = @import("std");
const z = @import("bar");
const a = @import("foo");
```

After

```zig
const z = @import("bar");
const a = @import("foo");
const std = @import("std");
```

But maybe special case `std` to the top?

```zig
const std = @import("std");
const z = @import("bar");
const a = @import("foo");
```

## Sorting by file import name

```zig
const baz = @import("baz.zig");
const bar = @import("bar.zig");
const foo = @import("foo.zig");
```

After

```zig
const bar = @import("bar.zig");
const baz = @import("baz.zig");
const foo = @import("foo.zig");
```

## Separating and inserting a blank line between package and file imports

Before

```zig
const std = @import("std");
const foo = @import("foo.zig");
const baz = @import("baz");
const bar = @import("bar.zig");
```

After

```zig
const std = @import("std");
const baz = @import("baz");

const foo = @import("foo.zig");
const bar = @import("bar.zig");
```
